### PR TITLE
Add get friend list endpoint

### PR DESCRIPTION
**Pull request by majorro**
Part of #135 

Complementing #530 adds an endpoint for fetching all player's friends

### CODE REVIEW

**Comment by majorro** [comment]
Pointing this to [feature/friends-list-storing](https://github.com/osu-tournament-rating/otr-api/tree/feature/friends-list-storing) becuase this pr would get flooded with that branch commits

File: API/Controllers/PlayersController.cs [hunk 1]
```diff
@@ -96,4 +96,22 @@ public async Task<IActionResult> ListAdminNotesAsync(int id)
 
         return Ok(await adminNoteService.ListAsync<PlayerAdminNote>(id));
     }
+
+    [HttpGet("{id:int}/friends")]
```
Comment (hburn7) [inline API/Controllers/PlayersController.cs:100]: Needs endpoint documentation, please see the other endpoints for examples.

File: API/Controllers/PlayersController.cs [hunk 2]
```diff
@@ -96,4 +96,22 @@ public async Task<IActionResult> ListAdminNotesAsync(int id)
...
+        {
+            return NotFound();
+        }
+
+        IEnumerable<PlayerCompactDTO>? result = await playerService.GetFriendsAsync(id);
+
+        return result is null
+            ? NotFound()
+            : Ok(result);
+    }
```
Comment (hburn7) [inline API/Controllers/PlayersController.cs:116]: I think this needs to live under the `UsersController` actually. Friends are only populated when a `User` interacts with the website. Thoughts? @myssto

File: API/Services/Implementations/PlayerService.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 namespace API.Services.Implementations;
 
-public class PlayerService(IPlayersRepository playerRepository, IMapper mapper) : IPlayerService
+public class PlayerService(IPlayersRepository playerRepository, IUserService userService, IMapper mapper) : IPlayerService
```
Comment (hburn7) [inline API/Services/Implementations/PlayerService.cs:9]: Assuming the endpoint is supposed to be on the `UsersController`, the changes in this file would be discarded.

**Review by hburn7 — CHANGES_REQUESTED** [review]
My thoughts so far. I think this will require a lot of changes if this is to be a user-derived feature rather than a player-derived one.

File: API/Controllers/PlayersController.cs (hunk 2 above)
Comment (majorro) [reply API/Controllers/PlayersController.cs:116]: I relied on the issue description, so I've done it this way. Imo it should be sticked to users too

Comment (hburn7) [reply API/Controllers/PlayersController.cs:116]: Yeah that issue is quite old, sorry.

File: API/Controllers/PlayersController.cs (hunk 1 above)
Comment (myssto) [reply API/Controllers/PlayersController.cs:100]: +1 Makes me wish I would stop putting off writing that docs page

File: API/Controllers/PlayersController.cs (hunk 2 above)
Comment (myssto) [reply API/Controllers/PlayersController.cs:116]: Yes, this endpoint should be on the `UsersController` instead

File: API/Controllers/PlayersController.cs [hunk 4]
```diff
@@ -96,4 +96,22 @@ public async Task<IActionResult> ListAdminNotesAsync(int id)
 
         return Ok(await adminNoteService.ListAsync<PlayerAdminNote>(id));
     }
+
+    [HttpGet("{id:int}/friends")]
+    [Authorize(Roles = $"{OtrClaims.Roles.User}, {OtrClaims.Roles.Client}")]
```
Comment (myssto) [inline API/Controllers/PlayersController.cs:101]:
```suggestion
    [Authorize(Policy = AuthorizationPolicies.AccessUserResources]
```
I think to keep the expectation of privacy the same across data sources this needs to be only readable by the resource owner. You are not implicitly able to read the friends lists of others on osu! which is where this data is coming from in the first place.

**Review by myssto — CHANGES_REQUESTED** [review]

File: API/Controllers/PlayersController.cs (hunk 1 above)
Comment (majorro) [reply API/Controllers/PlayersController.cs]: should've added a todo xd

File: API/Controllers/PlayersController.cs (hunk 2 above)
Comment (majorro) [reply API/Controllers/PlayersController.cs]: moved

File: API/Services/Interfaces/IUserService.cs [hunk 5]
```diff
@@ -33,6 +33,13 @@ public interface IUserService
     /// <returns>A list of match submissions, or null if not found</returns>
     Task<IEnumerable<MatchSubmissionStatusDTO>?> GetSubmissionsAsync(int id);
 
+    /// <summary>
+    /// Gets a user's friends for the given id
+    /// </summary>
+    /// <param name="id">Id of the user</param>
+    /// <returns>A list of user's friends</returns>
```
Comment (hburn7) [inline API/Services/Interfaces/IUserService.cs:40]:
```suggestion
    /// <returns>A list of the user's friends</returns>
```

**Review by hburn7 — CHANGES_REQUESTED** [review]
A lot better, just some minor things.

File: API/Controllers/UsersController.cs [hunk 6]
```diff
@@ -252,4 +252,23 @@ public async Task<IActionResult> SyncRulesetAsync(int id)
...
             : BadRequest();
     }
+
+    /// <summary>
+    /// Get a user's friends
+    /// </summary>
+    /// <param name="id">User id</param>
+    [HttpGet("{id:int}/friends")]
+    [Authorize(Policy = AuthorizationPolicies.AccessUserResources)]
+    [ProducesResponseType<IEnumerable<PlayerCompactDTO>>(StatusCodes.Status200OK)]
```
[the remark this answers is not in the document]
Comment (majorro) [reply API/Controllers/UsersController.cs:262]: I just want to ask myself

How in the world am I missed this???

Will fix later

**Comment by hburn7** [comment]
Going to go ahead and merge this into the base friends list branch but will be parking that feature to focus on more important items. Thanks for working on this one @majorro :)

**Pull request merged**